misc-tools/configure-domjudge

```diff
@@ -0,0 +1,137 @@
+i.e. use 8.1 for DOMjudge 8.1.x)
+
+Part of the DOMjudge Programming Contest Jury System and licensed
+under the GNU GPL. See README and COPYING for details.
+'''
+
+import fileinput
+import json
+import os.path
+import requests
+import requests.utils
+import sys
+from typing import List, Set
+import yaml
+
+headers = {
+    'user-agent': f'configure-domjudge ({requests.utils.default_user_agent()})'}
+
+
+def confirm(message: str, default: bool) -> bool:
+    answer = 'x'
+    while answer not in ['y', 'n']:
+        yn = 'Y/n' if default else 'y/N'
+        answer = input(f'{message} ({yn}) ').lower()
+        if answer == '':
+            answer = 'y' if default else 'n'
+    return answer == 'y'
+
+
+def usage():
+    print(f'Usage {sys.argv[0]} <domjudge-api-url>')
+    exit(1)
+
+
+def parse_api_response(name: str, response: requests.Response):
+    # The connection worked, but we may have received an HTTP error
+    if response.status_code >= 300:
+        print(response.text)
+        if response.status_code == 401:
+            raise RuntimeError(
+                'Authentication failed, please check your DOMjudge credentials in ~/.netrc.')
+        else:
+            raise RuntimeError(
+                f'API request {name} failed (code {response.status_code}).')
+
+    # We got a successful HTTP response. It worked. Return the full response
+    return json.loads(response.text)
+
+
+def do_api_request(name: str):
+    '''Perform an API call to the given endpoint and return its data.
+
+    Parameters:
+        name (str): the endpoint to call
+
+    Returns:
+        The endpoint contents.
+
+    Raises:
+        RuntimeError when the response is not JSON or the HTTP status code is non 2xx.
+    '''
+
+    url = f'{api_url}/{name}'
+
+    try:
+        response = requests.get(url, headers=headers)
+    except requests.exceptions.RequestException as e:
+        raise RuntimeError(e)
+
+    return parse_api_response(name, response)
+
+
+def compare_configs(expected_config: Set, actual_config: Set) -> (List, Set, Set):
+        diffs = []
+        for k in expected_config.keys():
+            if k in actual_config and expected_config[k] != actual_config[k]:
+                diffs.append(f'- {k}:\n   is: {actual_config[k]}\n  new: {expected_config[k]}')
+        if diffs:
+            print('Differences between actual and new configuration:')
+            for d in diffs:
+                print(d)
+
+        new_keys = set(expected_config.keys()).difference(set(actual_config.keys()))
+        missing_keys = set(actual_config.keys()).difference(set(expected_config.keys()))
+        return (diffs, new_keys, missing_keys)
+
+
+if len(sys.argv) < 2:
+    usage()
+
+api_url = sys.argv[1]
+
+user_data = do_api_request('user')
+if 'admin' not in user_data['roles']:
+    print('Your user does not have the \'admin\' role, can not import.')
+    exit(1)
+
+if os.path.exists('config.json'):
```

Comment:
Could we (also) do yaml? Maybe in the future

Reply:
Yes, should be trivial.

I started with JSON since it's easy to retrieve (and change locally to test with).

Reply:
👍